app/module/configurator.go

```diff
@@ -1,49 +1,83 @@
 package module
 
 import (
+	"context"
 	"fmt"
 
-	"github.com/gogo/protobuf/grpc"
+	pbgrpc "github.com/gogo/protobuf/grpc"
 
 	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	"github.com/cosmos/cosmos-sdk/types/module"
+	"google.golang.org/grpc"
 )
 
-type configurator struct {
-	cdc         codec.Codec
-	msgServer   grpc.Server
-	queryServer grpc.Server
+// VersionedConfigurator is a struct used at startup to register all the message and
+// query servers for all modules. It allows the module to register any migrations from
+// one consensus version of the module to the next. Finally it maps all the messages
+// to the app versions that they are accepted in. This then gets used in the antehandler
+// to prevent users from submitting messages that can not yet be executed.
+type VersionedConfigurator struct {
+	fromVersion, toVersion uint64
+	cdc                    codec.Codec
+	msgServer              pbgrpc.Server
+	queryServer            pbgrpc.Server
+	acceptedMessages       map[uint64]map[string]struct{}
 
 	// migrations is a map of moduleName -> fromVersion -> migration script handler
 	migrations map[string]map[uint64]module.MigrationHandler
 }
 
 // NewConfigurator returns a new Configurator instance
-func NewConfigurator(cdc codec.Codec, msgServer grpc.Server, queryServer grpc.Server) module.Configurator {
-	return configurator{
-		cdc:         cdc,
-		msgServer:   msgServer,
-		queryServer: queryServer,
-		migrations:  map[string]map[uint64]module.MigrationHandler{},
+func NewConfigurator(cdc codec.Codec, msgServer, queryServer pbgrpc.Server) VersionedConfigurator {
+	return VersionedConfigurator{
+		cdc:              cdc,
+		msgServer:        msgServer,
+		queryServer:      queryServer,
+		migrations:       map[string]map[uint64]module.MigrationHandler{},
+		acceptedMessages: map[uint64]map[string]struct{}{},
 	}
 }
 
-var _ module.Configurator = configurator{}
+var _ module.Configurator = VersionedConfigurator{}
+
+func (c *VersionedConfigurator) WithVersions(fromVersion, toVersion uint64) module.Configurator {
+	c.fromVersion = fromVersion
+	c.toVersion = toVersion
+	return c
+}
 
 // MsgServer implements the Configurator.MsgServer method
-func (c configurator) MsgServer() grpc.Server {
-	return c.msgServer
+func (c VersionedConfigurator) MsgServer() pbgrpc.Server {
+	return &serverWrapper{
+		addMessages: c.addMessages,
+		msgServer:   c.msgServer,
+	}
+}
+
+func (c VersionedConfigurator) addMessages(msgs []string) {
+	for version := c.fromVersion; version <= c.toVersion; version++ {
+		if _, exists := c.acceptedMessages[version]; !exists {
+			c.acceptedMessages[version] = map[string]struct{}{}
+		}
+		for _, msg := range msgs {
+			c.acceptedMessages[version][msg] = struct{}{}
```

Comment:
I actually don't know. It's how I've seen other engineers do it so I just copied it here but I never really benchmarked

Reply:
Turns out `struct{}` is more efficient: https://itnext.io/set-in-go-map-bool-and-map-struct-performance-comparison-5315b4b107b

Ppl argue back and forth on [Reddit](https://www.reddit.com/r/golang/comments/gnge5m/stop_using_mapstringstruct_for_existence_checks/) about it. My takeaways is `struct{}` is a bit worse syntax wise but better because the map can only be used for existence checks. If a `bool` is used, at some point later a developer could interpret `bool`s in the map for something else besides just existence.